### PR TITLE
[PERF] Parallelize reduction for CPU

### PR DESCRIPTION
**Pull request by icemelon9**
Can significantly reduce the latency for ops like mean and variance, but also increase latency for small reduction ops such as sum, max, etc. However these small ops are fast anyway. So I think it's worth to parallelize the reduction by default.

Will add some benchmark result soon...

@anijain2305 @Huyuwei @kevinthesun

### CODE REVIEW

**Comment by anijain2305** [comment]
Just to add, this is essential for QNN. QNN adds many `sum` operators on large tensors. Using this, I can cut down e2e latency by 40-50%.

File: topi/python/topi/x86/reduction.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                if tensor.op not in scheduled_ops:
+                    traverse_before_reduce(tensor.op)
+        else:
+            raise RuntimeError("Unsupported operator: %s" % operator.tag)
+
+        scheduled_ops.append(operator)
+
+    traverse_after_reduce(outs[0].op)
+    return sch
+
```
Comment (anijain2305) [inline topi/python/topi/x86/reduction.py]: Remove

File: topi/python/topi/x86/reduction.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+# pylint: disable=invalid-name
+"""x86 declaration and schedules."""
+from __future__ import absolute_import as _abs
+import tvm
+from .. import tag
+from .. import generic
+
+
+def _schedule_reduce(sch, out, is_idx_reduce=False):
+    if len(sch[out].op.axis) >= 5:
```
Comment (soiferj) [inline topi/python/topi/x86/reduction.py]: These two blocks are the same - is that intentional? Is there a reason to not always fuse all outer axes?

File: topi/python/topi/x86/reduction.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+# under the License.
+# pylint: disable=invalid-name
+"""x86 declaration and schedules."""
+from __future__ import absolute_import as _abs
+import tvm
+from .. import tag
+from .. import generic
+
+
+def _schedule_reduce(sch, out, is_idx_reduce=False):
```
Comment (kevinthesun) [inline topi/python/topi/x86/reduction.py]: is_idx_reduce not used?

**Comment by anijain2305** [comment]
@icemelon9 Did you get a chance to work on the comments?

**Comment by tqchen** [comment]
ping @icemelon9

**Comment by anijain2305** [comment]
Ping @icemelon9 :)

File: topi/python/topi/x86/reduction.py (hunk 3 above)
Comment (icemelon9) [reply topi/python/topi/x86/reduction.py]: fixed

File: topi/python/topi/x86/reduction.py (hunk 1 above)
Comment (icemelon9) [reply topi/python/topi/x86/reduction.py]: this is necessary

**Comment by icemelon9** [comment]
ci is green now. @anijain2305 @kevinthesun @soiferj Could you review this again?

**Comment by yzhliu** [comment]
Thanks @icemelon9 @anijain2305 @soiferj @kevinthesun

**Pull request merged**